api/http/management_filetransfer.go

```diff
@@ -0,0 +1,736 @@
+	sessionTopic := model.GetSessionSubject(params.TenantID, params.SessionID)
+	msgChan := make(chan *natsio.Msg, channelSize)
+	sub, err := h.nats.ChanSubscribe(sessionTopic, msgChan)
+	if err != nil {
+		responseError = errors.Wrap(err, errFileTranserSubscribing.Error())
+		return
+	}
+
+	//nolint:errcheck
+	defer sub.Unsubscribe()
+
+	// stat the remote file
+	req := wsft.StatFile{
+		Path: request.Path,
+	}
+	if err := h.publishFileTransferProtoMessage(params.SessionID,
+		params.UserID, deviceTopic, wsft.MessageTypeStat, req, 0); err != nil {
+		responseError = err
+		return
+	}
+
+	ticker := time.NewTicker(fileTransferPingInterval)
+	defer ticker.Stop()
+
+	// handle messages from the device
+	timeout := time.NewTimer(fileTransferTimeout)
+	latestOffset := int64(0)
+	numberOfChunks := 0
+	for {
+		select {
+		case wsMessage := <-msgChan:
+			msg, msgBody, err := h.decodeFileTransferProtoMessage(wsMessage.Data)
+			if err != nil {
+				responseError = err
+				return
+			}
+
+			// reset the timeout ticket
+			timeout.Reset(fileTransferTimeout)
+
+			// process incoming messages from the device by type
+			switch msg.Header.MsgType {
+
+			// error message, stop here
+			case wsft.MessageTypeError:
+				errorMsg := msgBody.(*wsft.Error)
+				if *errorMsg.MessageType == wsft.MessageTypeStat {
+					responseError = errors.Wrap(errors.New(*errorMsg.Error),
+						errFileTranserFailed.Error())
+				} else {
+					responseError = errors.New(*errorMsg.Error)
+				}
+				return
+
+			// file stat response, if okay, let's get the file
+			case wsft.MessageTypeFileInfo:
+				req := wsft.GetFile{
+					Path: request.Path,
+				}
+				if err := h.publishFileTransferProtoMessage(params.SessionID,
+					params.UserID, deviceTopic, wsft.MessageTypeGet,
+					req, 0); err != nil {
+					responseError = err
+					return
+				}
+
+				fileInfo := msgBody.(*wsft.FileInfo)
+				writeHeaders(c, fileInfo)
+				responseHeaderSent = true
+
+			// file data chunk
+			case wsft.MessageTypeChunk:
+				if msg.Body == nil {
+					if err := h.publishFileTransferProtoMessage(
+						params.SessionID, params.UserID, deviceTopic,
+						wsft.MessageTypeACK, nil,
+						latestOffset); err != nil {
+						return
+					}
+					return
+				}
+
+				// verify the offset property
+				propOffset, _ := msg.Header.Properties[PropertyOffset].(int64)
+				if propOffset != latestOffset {
+					responseError = errors.Wrap(errFileTranserFailed,
+						"wrong offset received")
+					return
+				}
+				latestOffset += int64(len(msg.Body))
+
+				_, err := c.Writer.Write(msg.Body)
+				if err != nil {
+					return
+				}
+
+				numberOfChunks++
+				if numberOfChunks >= ackSlidingWindowSend {
+					if err := h.publishFileTransferProtoMessage(
+						params.SessionID, params.UserID, deviceTopic,
+						wsft.MessageTypeACK, nil,
+						latestOffset); err != nil {
+						return
+					}
+					numberOfChunks = 0
+				}
+			}
+
```

Comment:
```suggestion

			case ws.MessageTypePing:
				// send pong
				if err := h.publishFileTransferProtoMessage(
						params.SessionID, params.UserID, deviceTopic,
						ws.MessageTypePong, nil,
						-1); err != nil {
						return
					}
				fallthrough
			case ws.MessageTypePong:
				ticker.Reset(fileTransferPingInterval)
```
```

Reply:
I didn't know the existence of the `fallthrough` keyword!